File: src/java/billiards/viewer/SideSum.java
package billiards.viewer;

public final class SideSum {

    private final int[] coeffs;

    private final double x;
    private final double y;
    private final double z;

    private SideSum(final int[] coeffs, final double x, final double y, final double z) {
    	this.coeffs = coeffs;
    	this.x = x;
    	this.y = y;
    	this.z = z;

    }
    
    public static SideSum create(final double x, final double y) {
    	
    	return new SideSum( new int[3], x, y, Math.PI - x - y);
    }

    public void add(final int num) {
        this.coeffs[num] += 1;
    }

    public void sub(final int num) {
        this.coeffs[num] -= 1;
    }

    public double sum() {
        return this.coeffs[0] * this.x + this.coeffs[1] * this.y + this.coeffs[2] * this.z;
    }
    
    public SideSum copy() {
    	final int[] newCoeffs = new int[coeffs.length];
    	for (int i = 0; i < coeffs.length; i++) {
    		final int coef = coeffs[i];
    		newCoeffs[i] = coef;
    	}
    	return new SideSum(newCoeffs, this.x, this.y, this.z);
    }
    
    @Override
    public String toString() {
    	return this.coeffs[0] + " * A + " + this.coeffs[1] 
    			+ " * B + " + this.coeffs[2] + " * C";
    }
}

File: src/backend/cpp/math/side_sum.cpp
#include "math/side_sum.hpp"

/*
Jul 31 2025 Marco Mai
transfer from Java
*/
SideSum::SideSum(const std::array<int32_t, 3>& coeffs,const float64_t& x,const float64_t& y,const float64_t& z)
    : coeffs(coeffs), x(x), y(y), z(z) {}



SideSum SideSum::create(const float64_t& x, const float64_t& y) {
    return SideSum({0, 0, 0}, x, y, boost::math::constants::pi<float64_t>() - x - y);
}


void SideSum::add(int32_t num) {
    coeffs[num] += 1;
}

void SideSum::sub(int32_t num) {
    coeffs[num] -= 1;
}

float64_t SideSum::sum() {
    return coeffs[0] * x + coeffs[1] * y + coeffs[2] * z;
}


SideSum SideSum::copy() const {
    return SideSum(coeffs, x, y, z);
}

std::string SideSum::to_string() const {
    std::ostringstream oss;
    oss << coeffs[0] << " * A + " << coeffs[1]
        << " * B + " << coeffs[2] << " * C";
    return oss.str();
}
